pkg/controller/enterprisesearch/config.go

```diff
@@ -224,43 +224,51 @@ func configRefWatchName(ent types.NamespacedName) string {
 	return fmt.Sprintf("%s-%s-configref", ent.Namespace, ent.Name)
 }
 
-// parseConfigRef builds a single merged CanonicalConfig from the secrets referenced in configRef,
-// and ensures watches are correctly set on those secrets.
+// parseConfigRef builds a CanonicalConfig from the secret referenced in configRef,
+// and ensures a watch is set on this secret.
 func parseConfigRef(driver driver.Interface, ent entv1beta1.EnterpriseSearch) (*settings.CanonicalConfig, error) {
-	cfg := settings.NewCanonicalConfig()
-	secretNames := make([]string, 0, len(ent.Spec.ConfigRef))
-	for _, secretRef := range ent.Spec.ConfigRef {
-		if secretRef.SecretName == "" {
-			continue
-		}
-		secretNames = append(secretNames, secretRef.SecretName)
+	var secretNames []string
+	if ent.Spec.ConfigRef != nil && ent.Spec.ConfigRef.SecretName != "" {
+		secretNames = append(secretNames, ent.Spec.ConfigRef.SecretName)
 	}
 	nsn := k8s.ExtractNamespacedName(&ent)
+	// will reset any existing watch if secretNames is empty
```

Comment:
Looks right but seems like a side-effect of the function IMO.
I noticed that we are doing almost the same thing for beats. Could we factorize the code?
I also see that for beats we [return early](https://github.com/elastic/cloud-on-k8s/blob/29d7abca1e4a8f60f8712b2f64814d9aa51d489e/pkg/controller/beat/common/config.go#L117-L138) so it looks like we could leak a watch.

Reply:
Yes! I think it will be easy enough to factorize, and fix the Beat watch leak at the same time 👍